### PR TITLE
Update lanes to build stable and beta releases

### PR DESCRIPTION
**Pull request by loremattei**
This PR updates `build_and_upload_release` so that it doesn't expect the tag to be already present in the repository. 
The user can now switch between beta and stable builds passing the right value to the new `beta_release` parameter. 

When `create_gh_release` is passed with a `true` value, the lane will create the draft of the release on GitHub associating the HEAD commit of the branch to it. When the release manager releases the draft, GitHub creates the actual tag in the repository. This means that we don't need to give more access right to the CI to be able to push the tag as we get the tag "for free" with out current process. 

The PR also adds two new lanes `build_and_upload_beta_release` and `build_and_upload_stable_release` which invoke `build_and_upload_release` with the correct value of the `beta_release` flag.

Also, this PR updates `Fastlane` to the latest. 

A follow up PR will take care of updating the CI configuration and the related Fastlane lanes.

**To test:**
1. Create a new test branch out of this one. 
2. Update the Internal and Public versions in the xcconfig files. Please, use a version that is not used in production (anything with a minor number > 9 will work).
3. Comment out the code to upload [to iTC](https://github.com/wordpress-mobile/WordPress-iOS/blob/944ac9db29e6be4c5f0519195b3f192bb8b2edc0/fastlane/Fastfile#L440) and [to AppCenter](https://github.com/wordpress-mobile/WordPress-iOS/blob/944ac9db29e6be4c5f0519195b3f192bb8b2edc0/fastlane/Fastfile#L391). These changes don't affect that part of the code, so we don't need to upload dummy stuff, especially to TestFlight where it can be removed. 
4. Run `bundle exec fastlane build_and_upload_beta_release create_gh_release:true` and verify that the build succeeds and that releases are drafted in GitHub for the beta (_Note_: don't release it or a tag will be added and the build started on CI and uploaded)
5. Run `bundle exec fastlane build_and_upload_stable_release create_gh_release:true` and verify that the release is drafted for stable  (_Note_: don't release it or a tag will be added and the build started on CI and uploaded).
6. Cleanup all the dummy releases on GitHub.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by AliSoftware** [comment]
First skim at the PR code (not tested yet), and I wonder if it won't start to get confusing to have both `build_and_upload_internal` and `build_and_upload_beta_release` as well as having both `build_and_upload_itc` and `build_and_upload_stable_release`… I mean, without looking at the implementation details for each lane and just looking at the lane names, it was not clear to me at first glance just from naming what was the difference between the ones in each pair.

Note that it might just be because I'm not used to WP release process and kinda forgot that we do internal betas via appcenter, so my first read/impression made me confuse "internal vs testflight" as if it was the same as "beta vs stable", while after a closer look I realized those are orthogonal stuff (both beta and stable each trigger both internal+testflight, right?)… so now, if I look closer I can see that we have the following (see below)… while it wasn't obvious to me at first, again probably because I'm not used to doing internal builds when I deal with WC.

```
┌────────────────────────────────┐  ┌────────────────────────────────┐
│ build_and_upload_beta_release  │  │build_and_upload_stable_release │
└────────────────────────────────┘  └────────────────────────────────┘
                 │                                   │
                 └─────────────────┬─────────────────┘
                                   │
                                   ▼
                  ┌────────────────────────────────┐
                  │  build_and_upload_release(…)   │
                  └────────────────────────────────┘
                                   │
                 ┌─────────────────┴─────────────────┐
                 │                                   │
                 ▼                                   ▼
┌────────────────────────────────┐  ┌────────────────────────────────┐
│   build_and_upload_internal    │  │      build_and_upload_itc      │
└────────────────────────────────┘  └────────────────────────────────┘
                 │                                   │
                 ▼                                   ▼
┌────────────────────────────────┐  ┌────────────────────────────────┐
│gym(…), appcenter(…), sentry(…) │  │gym(…), testflight(…), sentry(…)│
└────────────────────────────────┘  └────────────────────────────────┘
```
I mean, once I see that dependency graph it's suddenly all more clear to me, but without it and with just seeing the names (e.g. `fastlane lanes` to print existing lanes), it's not that obvious. I wonder if we couldn't find more descriptive names… though thinking about it I don't have any real suggestion in practice, so just dropping that here to know what you think and to give feedback from the perspective of someone not used to WP's RM.

---

_Note: I plan to do a proper review by testing the changes in more details later today._

File: fastlane/Fastfile [hunk 1]
```diff
@@ -257,32 +257,74 @@ import "./ScreenshotFastfile"
...
+  # Example:
+  # bundle exec fastlane build_and_upload_beta_release
+  # bundle exec fastlane build_and_upload_beta_release skip_confirm:true
+  # bundle exec fastlane build_and_upload_beta_release create_gh_release:true
+  #####################################################################################
+  desc "Builds and uploads a beta release for distribution"
+  lane :build_and_upload_beta_release do | options |
+    build_and_upload_release(skip_confirm: options[:skip_confirm], 
+      create_gh_release: options[:create_gh_release],
+      beta_release: true)
```
Comment (AliSoftware) [inline fastlane/Fastfile]: nit: formatting
```suggestion
    build_and_upload_release(
      skip_confirm: options[:skip_confirm], 
      create_gh_release: options[:create_gh_release],
      beta_release: true
    )
```

File: fastlane/Fastfile [hunk 2]
```diff
@@ -257,32 +257,74 @@ import "./ScreenshotFastfile"
...
+  # Example:
+  # bundle exec fastlane build_and_upload_stable_release
+  # bundle exec fastlane build_and_upload_stable_release skip_confirm:true
+  # bundle exec fastlane build_and_upload_stable_release create_gh_release:true
+  #####################################################################################
+  desc "Builds and uploads a stable release for distribution"
+  lane :build_and_upload_stable_release do | options |
+    build_and_upload_release(skip_confirm: options[:skip_confirm], 
+      create_gh_release: options[:create_gh_release],
+      beta_release: false)
```
Comment (AliSoftware) [inline fastlane/Fastfile]: nit: formatting
```suggestion
    build_and_upload_release(
      skip_confirm: options[:skip_confirm], 
      create_gh_release: options[:create_gh_release],
      beta_release: false
    )
```

**Comment by AliSoftware** [comment]
> 4. Run `bundle exec fastlane build_and_upload_beta_release create_gh_release:true` and verify that the build succeeds and that **releases are drafted in GitHub for both beta and internal**

That is not exactly what happens though 🤔  but I think the behavior is correct but the description misleading?

Instead, the lane builds both internal and itc, but one GitHub release gets drafted at the end – the one for iTC, with the `xcarchive` that is signed for AppStore cert as its only asset.

<details><summary>fastlane summary – list of lanes called</summary>

```
+------+------------------------------------------+-------------+
|                       fastlane summary                        |
+------+------------------------------------------+-------------+
| Step | Action                                   | Time (in s) |
+------+------------------------------------------+-------------+
| 1    | default_platform                         | 0           |
| 2    | default_platform                         | 0           |
| 3    | is_ci                                    | 0           |
| 4    | setup_circle_ci                          | 0           |
| 5    | Switch to build_and_upload_release lane  | 0           |
| 6    | ios_build_prechecks                      | 19          |
| 7    | ios_build_preflight                      | 36          |
| 8    | Switch to build_and_upload_internal lane | 0           |
| 9    | Switch to internal_code_signing lane     | 0           |
| 10   | match                                    | 33          |
| 11   | gym                                      | 460         |
| 12   | sentry_upload_dsym                       | 17          |
| 13   | Switch to build_and_upload_itc lane      | 0           |
| 14   | Switch to appstore_code_signing lane     | 0           |
| 15   | match                                    | 22          |
| 16   | gym                                      | 401         |
| 17   | sentry_upload_dsym                       | 17          |
| 18   | zip                                      | 16          |
| 19   | ios_get_app_version                      | 0           |
| 20   | create_release                           | 43          |
+------+------------------------------------------+-------------+
```
</details>
<details><summary>GH release after <tt>build_and_upload_stable_release create_gh_release:true</tt></summary>
build_and_upload_beta_release

Extract of the `Info.plist` from within the `xcarchive` attached to that GH release:
```xml
<key>CFBundleVersion</key>
		<string>16.15918.0.100</string>
		<key>SigningIdentity</key>
		<string>Apple Distribution: Automattic, Inc. (PZYM8XX95Q)</string>
		<key>Team</key>
		<string>PZYM8XX95Q</string>
```
</details>

That behavior matches the code in the Fastlane though, since the `build_and_upload_internal` lane indeed doesn't at any point call `create_gh_release()` nor use that option, and that PR didn't change that… so I think that's expected, despite what the PR description says?

_PS: Running the lane for stable now, will report back on that one once it has finished ~reheating my room~ building_ 😅

**Comment by AliSoftware** [comment]
PS / Note for future reviewers: I had to skip all the files that `configure_apply` was suggesting me to override when I was trying the lanes, since the branch of this PR hasn't been updated with the latest secrets (so I needed it to keep using the ones I applied when I was on develop instead)

**Comment by AliSoftware** [comment]
> 5. Run `bundle exec fastlane build_and_upload_stable_release create_gh_release:true` and verify that the release is drafted for stable (Note: don't release them or a tag will be added and the build started on CI and uploaded).

That one works as described, i.e. only builds itc, and creates the same GitHub Release draft as before (single GH release, with a single asset being the `xcarchive` signed for AppStore), so all good  👍 

<details><summary>fastlane summary – list of lanes called</summary>

```
+------+-----------------------------------------+-------------+
|                       fastlane summary                       |
+------+-----------------------------------------+-------------+
| Step | Action                                  | Time (in s) |
+------+-----------------------------------------+-------------+
| 1    | default_platform                        | 0           |
| 2    | default_platform                        | 0           |
| 3    | is_ci                                   | 0           |
| 4    | setup_circle_ci                         | 0           |
| 5    | Switch to build_and_upload_release lane | 0           |
| 6    | ios_build_prechecks                     | 7           |
| 7    | ios_build_preflight                     | 34          |
| 8    | Switch to build_and_upload_itc lane     | 0           |
| 9    | Switch to appstore_code_signing lane    | 0           |
| 10   | match                                   | 30          |
| 11   | gym                                     | 414         |
| 12   | sentry_upload_dsym                      | 15          |
| 13   | zip                                     | 17          |
| 14   | ios_get_app_version                     | 0           |
| 15   | create_release                          | 28          |
+------+-----------------------------------------+-------------+
```
</details>

<details><summary>Draft GitHub Release created after <tt>build_and_upload_stable_release create_gh_release:true</tt></summary>

</details>

**Review by AliSoftware — COMMENTED** [review]
Ok, see comments for details but I think this works as expected. Will just leave you confirm that the behavior for `build_and_upload_beta_release create_gh_release:true` is expected, despite not matching the description.

Apart from that question, LGTM! 👍

**Comment by loremattei** [comment]
Thank you @AliSoftware! You're right about the internal release not being created on GitHub. I've update the PR description.
Also, I've updated the code to address your comments.

**Review by AliSoftware — APPROVED** [review]
All good then, :shipit: Merge on 🟢

**Pull request merged**